### PR TITLE
Credit Card Feature

### PR DESCRIPTION
**Pull request by christian-huntington**
Refactoring:
- Many classes/methods were refactored according the design patterns discussed at last meeting.

Implementation of external Plugin (PDF / Printing):
- Included 2 external libs on the pubspec.yaml (pdf and printing)
- Those libs are used to generate a Payment Confirmation PDF file and to allow the user to print/share that file.

Mockoon JSON file:
- Included Credit Card paths and json responses. 

Unit tests:
- More tests were implemented.

### CODE REVIEW

**Comment by aelshamy** [comment]
@christian-huntington can you fix the conflicts

**Comment by christian-huntington** [comment]
@aelshamy I have resolve the conflicts related to credit_card feature, but I removed any changes to the mockoon json file. Plz merge these changes and I'll work on that file soon.

File: business_banking/test/features/credit_card/ui/payment_request/credit_card_payemnt_request_screen_test.dart [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  tearDown(() {
+    viewModel = null;
+    actions = null;
+    testWidget = null;
+  });
+
+  testWidgets('CreditCardPaymentRequestScreen initialization', (tester) async {
+    await tester.pumpWidget(testWidget);
+    await tester.pump(Duration(milliseconds: 500));
+    expect(find.byType(CreditCardPaymentRequestScreen), findsOneWidget);
```
Comment (aelshamy) [inline business_banking/test/features/credit_card/ui/payment_request/credit_card_payemnt_request_screen_test.dart:47]: if you are testing the screen, you need to verify the elements on the screen themselves

File: business_banking/test/features/credit_card/ui/payment_request/credit_card_payemnt_request_screen_test.dart [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  });
+
+  /*
+  testWidgets('CreditCardScreen viewModel values x screen values',
```
Comment (aelshamy) [inline business_banking/test/features/credit_card/ui/payment_request/credit_card_payemnt_request_screen_test.dart:51]: also remove the comments

**Review by aelshamy — CHANGES_REQUESTED** [review]

File: business_banking/test/features/credit_card/bloc/credit_card_bloc_test.dart [hunk 3]
```diff
@@ -31,4 +34,53 @@ void main() {
...
+      lastFour: '4444',
+      balance: 1234.56,
+      paymentDueDate: DateTime.parse("2021-05-03"),
+      nextClosingDate: DateTime.parse("2021-05-06"),
+      paymentMinimumValue: 100.00,
+      paymentValue: 20.00,
+    );
+    var event = CreditCardViewEventUpdatePaymentValue(viewModel, 0);
+    bloc.creditCardViewEventsPipeHandler(event);
+    verify(creditCardPaymentRequestUseCaseMock.updatePaymentValue(any)).called(1);
```
[the remark this answers is not in the document]
Comment (christian-huntington) [reply business_banking/test/features/credit_card/bloc/credit_card_bloc_test.dart:53]: OK! Got it! I've changed it to: 

verify(creditCardPaymentRequestUseCaseMock.updatePaymentValue(event.paymentValue)).called(1);

File: business_banking/lib/features/credit_card/bloc/credit_card_bloc.dart [hunk 4]
```diff
@@ -75,88 +75,69 @@ class CreditCardBloc extends Bloc {
...
-  final BuildContext context;
-
-  CreditCardViewEventNavigateBackToDetailsScreen(this.context);
-
-  @override
-  List<Object?> get props => [];
+/// view events
 
-}
+abstract class CreditCardViewEvents extends Equatable {}
```
[the remark this answers is not in the document]
Comment (christian-huntington) [reply business_banking/lib/features/credit_card/bloc/credit_card_bloc.dart:133]: OK! Moved to a new file called "bloc/credit_card_view_events.dart".

File: business_banking/test/features/credit_card/bloc/credit_card_bloc_test.dart [hunk 5]
```diff
@@ -31,4 +34,53 @@ void main() {
...
+      name: 'My Credit Card',
+      lastFour: '4444',
+      balance: 1234.56,
+      paymentDueDate: DateTime.parse("2021-05-03"),
+      nextClosingDate: DateTime.parse("2021-05-06"),
+      paymentMinimumValue: 100.00,
+      paymentValue: 20.00,
+    );
+    bloc.validatePaymentInformation(viewModel, 10);
+    verify(creditCardPaymentRequestUseCaseMock.validatePaymentInformation(any)).called(1);
```
[the remark this answers is not in the document]
Comment (christian-huntington) [reply business_banking/test/features/credit_card/bloc/credit_card_bloc_test.dart:70]: OK!

File: business_banking/test/features/credit_card/ui/credit_card/credit_card_actions_test.dart [hunk 6]
```diff
@@ -0,0 +1,54 @@
+// @dart=2.9
+
+import 'package:business_banking/features/credit_card/bloc/credit_card_bloc.dart';
+import 'package:flutter/material.dart';
+import 'package:mockito/mockito.dart';
+
+class CreditCardBlocMock extends Mock implements CreditCardBloc {}
+class MockBuildContext extends Mock implements BuildContext {}
+
+void main() {
```
[the remark this answers is not in the document]
Comment (christian-huntington) [reply business_banking/test/features/credit_card/ui/credit_card/credit_card_actions_test.dart:10]: I'm having problems when testing "navigation". I'm going to remove this file and check with you later how to test CFRouterScope navigation.

File: business_banking/test/features/credit_card/ui/payment_request/credit_card_payemnt_request_screen_test.dart (hunk 1 above)
Comment (christian-huntington) [reply business_banking/test/features/credit_card/ui/payment_request/credit_card_payemnt_request_screen_test.dart:47]: OK! I've added the tests to check the screen elements and event actions.

**Review by aelshamy — APPROVED** [review]

File: business_banking/lib/features/credit_card/bloc/credit_card_bloc.dart [hunk 7]
```diff
@@ -5,9 +5,9 @@ import 'package:business_banking/features/credit_card/model/payment_request/cred
 import 'package:business_banking/features/credit_card/model/payment_response/credit_card_payment_response_view_model.dart';
 import 'package:clean_framework/clean_framework.dart';
 import 'package:equatable/equatable.dart';
-import 'package:flutter/cupertino.dart';
-import 'package:flutter/material.dart';
-
+import 'package:intl/intl.dart';
+import 'package:pdf/pdf.dart';
+import 'package:pdf/widgets.dart' as pw;
```
[the remark this answers is not in the document]
Comment (aelshamy) [reply business_banking/lib/features/credit_card/bloc/credit_card_bloc.dart:10]: I thought you update that, right?

File: business_banking/lib/features/credit_card/bloc/credit_card_bloc.dart [hunk 8]
```diff
@@ -75,88 +75,69 @@ class CreditCardBloc extends Bloc {
...
-class CreditCardViewEventNavigateToPaymentRequestScreen extends CreditCardViewEvents {
-
-  final CreditCardViewModel viewModel;
-  final BuildContext context;
-
-  CreditCardViewEventNavigateToPaymentRequestScreen(this.viewModel, this.context);
-
-  @override
-  List<Object?> get props => [viewModel];
+  Future<pw.Document> generatePDFPaymentConfirmation(CreditCardPaymentResponseViewModel viewModel) async {
```
[the remark this answers is not in the document]
Comment (aelshamy) [reply business_banking/lib/features/credit_card/bloc/credit_card_bloc.dart:79]: did you push this update, I saw yesterday that you moved to the usecase, but it is not showing here

**Review by aelshamy — CHANGES_REQUESTED** [review]
please check the last two comments

**Pull request merged**